Import WiFi components from WiseConnect

jerome-pouiller opened the pull request:
- gecko: Import sl_bit.h and sl_string.h
- wiseconnect: Add WiFi components of WiSeConnect SDK
- wiseconnect: Apply changes to WiseConnect

jerome-pouiller committed:
wiseconnect: Add WiFi components of WiSeConnect SDK

Origin: Silicon Labs WiSeConnect SDK
License: MSLA
URL: https://github.com/siliconlabs/wiseconnect
Commit: e97a0ed00ddda347a8a39e8276f470e1c5fea469
Version: 3.3
Purpose: Add WiFi support for SiWx917

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
wiseconnect: Apply changes to WiseConnect

In order to compile, a few changes are needed. This patch mainly remove
orphan functions.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
gecko: Import sl_bit.h and sl_string.h

WiseConnect and GeckoSDK share some APIs. The upcoming WiseConnect
update will require sl_bit.h and sl_string.h.

Note the original path of these files is platform/common/inc/ while this
patch locate them in common/inc/.

Origin: Silicon Labs Gecko SDK
License: Zlib
URL: https://github.com/SiliconLabs/gecko_sdk
Commit: 124fa19de8c8b3961d21c20857f7df32239786da
Version: 4.4
Purpose: Required by WiseConnect

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
wiseconnect: Fix compiler warning

Compiler complain with:
    .../modules/hal/silabs/wiseconnect/components/device/silabs/si91x/wireless/asynchronous_socket/src/sl_si91x_socket.c: In function 'sli_si91x_accept_async':
    .../modules/hal/silabs/wiseconnect/components/device/silabs/si91x/wireless/asynchronous_socket/src/sl_si91x_socket.c:243:5: error: implicit declaration of function 'close'; did you mean 'fclose'? [-Werror=implicit-function-declaration]
      243 |     close(client_socket_id);
          |     ^~~~~
          |     fclose

close() is indeed declared in the BSD socket abstraction provided by
WiseConnect:

    int close(int socket_id)
    {
      errno = 0; // Clear any existing error
      return sli_si91x_shutdown(socket_id, SHUTDOWN_BY_ID);
    }

To avoid any confusion between BSD sockets and WiseConnect API, this
patch just call the underlying function (sli_si91x_shutdown()) instead
of close().

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jhedberg commented on wiseconnect/components/common/inc/sl_constants.h:
```diff
@@ -0,0 +1,184 @@
+/*
+ * EVALUATION AND USE OF THIS SOFTWARE IS SUBJECT TO THE TERMS AND
+ * CONDITIONS OF THE CONTROLLING LICENSE AGREEMENT FOUND AT LICENSE.md
+ * IN THIS SDK. IF YOU DO NOT AGREE TO THE LICENSE TERMS AND CONDITIONS,
+ * PLEASE RETURN ALL SOURCE FILES TO SILICON LABORATORIES.
+ * (c) Copyright 2018, Silicon Laboratories Inc.  All rights reserved.
+ */
```
This looks a different kind of license header than the Zlib and MSLA ones. We also don't have any `LICSENSE.md` file in the tree (or this PR).

jerome-pouiller replied:
I have added `license.md`.

jerome-pouiller commented:
v2:
  - import license.md, so header in `sl_constants.h` refers an existing file

asmellby commented on wiseconnect/components/device/silabs/si91x/mcu/core/cmsis/core_cm4.h:
Why do we need this file instead of using the separate CMSIS module from Zephyr?

jerome-pouiller commented:
v3:
  - remove useless core_cm4.h

jerome-pouiller committed:
wiseconnect: Fix memory access in sli_si91x_accept_async()

This issue has been reported to upstream and is fixed in WiseConnect
3.3.1

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
wiseconnect: Fix conflict with Zephyr definition of "fd_set"

Zephyr also export its version of fd_set. The two definitions are
incompatible.

This patch has been is mostly done by:

    sed -i 's/\bFD_/SL_SI91X_FD_/' **/*.[ch]

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller commented:
v4:
  - Import missing files for socket offloading
  - Fix `six917_accept()`
  - Allow to use WiseConnect `fd_set` implementation in parallel of the Zephyr one.

asmellby reviewed (approved):
We need to start the process of relicensing all the files we use to Zlib, but that shouldn't block merging into our fork for now.

jhedberg commented:
> We need to start the process of relicensing all the files we use to Zlib, but that shouldn't block merging into our fork for now.

Sure, that's fine. My biggest concern was the complete lack of license in one case, however that's now fixed.
